internal/confmapprovider/discovery/config.go

```diff
@@ -0,0 +1,356 @@
+	ErrorF(path string, err error) error
+	ToStringMap() map[string]any
+}
+
+type Entry map[string]any
+
+func (e Entry) ToStringMap() map[string]any {
+	return e
+}
+
+type ServiceEntry struct {
+	Entry `yaml:",inline"`
+}
+
+func (ServiceEntry) ErrorF(path string, err error) error {
+	return errorF(typeService, path, err)
+}
+
+type ExtensionEntry struct {
+	Entry `yaml:",inline"`
+}
+
+func (ExtensionEntry) ErrorF(path string, err error) error {
+	return errorF(typeExtension, path, err)
+}
+
+type ExporterEntry struct {
+	Entry `yaml:",inline"`
+}
+
+func (ExporterEntry) ErrorF(path string, err error) error {
+	return errorF(typeExporter, path, err)
+}
+
+type ObserverEntry struct {
+	Entry `yaml:",inline"`
+}
+
+func (ObserverEntry) ErrorF(path string, err error) error {
+	return errorF(typeDiscoveryObserver, path, err)
+}
+
+type ProcessorEntry struct {
+	Entry `yaml:",inline"`
+}
+
+func (ProcessorEntry) ErrorF(path string, err error) error {
+	return errorF(typeProcessor, path, err)
+}
+
+type ReceiverEntry struct {
+	Entry `yaml:",inline"`
+}
+
+func (ReceiverEntry) ErrorF(path string, err error) error {
+	return errorF(typeReceiver, path, err)
+}
+
+type ReceiverToDiscoverEntry struct {
+	// Receiver creator rules by observer extension ID
+	Rule map[config.ComponentID]string
+	// Platform/observer specific config by observer extension ID.
+	// These are merged w/ "default" component.ID in a "config" map
+	Config map[config.ComponentID]map[string]any
+	// The remaining items used to merge applicable rule and config
+	Entry `yaml:",inline"`
+}
+
+func (r ReceiverToDiscoverEntry) ToStringMap() map[string]any {
+	return r.Entry
+}
+
+func (ReceiverToDiscoverEntry) ErrorF(path string, err error) error {
+	return errorF(typeReceiverToDiscover, path, err)
+}
+
+func (c *Config) Load(discoveryDPath string) error {
+	if c == nil {
+		return fmt.Errorf("config must not be nil to be loaded (use NewConfig())")
+	}
+	return filepath.WalkDir(discoveryDPath, func(path string, d fs.DirEntry, err error) error {
+		c.logger.Debug(fmt.Sprintf("loading component: WalkDir(%s, %#v, %v)", path, d, err))
+		if err != nil {
+			return err
+		}
+		switch {
+		case isServiceEntryPath(path):
+			return loadEntry(typeService, path, c.Service)
+		case isExporterEntryPath(path):
+			return loadEntry(typeExporter, path, c.Exporters)
+		case isExtensionEntryPath(path):
+			if isDiscoveryObserverEntryPath(path) {
+				return loadEntry(typeDiscoveryObserver, path, c.DiscoveryObservers)
+			}
+			return loadEntry(typeExtension, path, c.Extensions)
+		case isProcessorEntryPath(path):
+			return loadEntry(typeProcessor, path, c.Processors)
+		case isReceiverEntryPath(path):
+			if isReceiverToDiscoverEntryPath(path) {
+				return loadEntry(typeReceiverToDiscover, path, c.ReceiversToDiscover)
+			}
+			return loadEntry(typeReceiver, path, c.Receivers)
+		default:
+			c.logger.Debug(fmt.Sprintf("Disregarding %s", path))
```

Comment:
Can we use zap fields?

Reply:
I'd prefer not to for a single line since the format is cumbersome imo, unless you think these should all be machine readable.

Reply:
in the past I've been told to avoid concatenations when logging and prefer to use late interpolation.

Reply:
that said this was in java, go might not be all that bad at it